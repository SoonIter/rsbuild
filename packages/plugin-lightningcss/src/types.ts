import type { CustomAtRules, TransformOptions } from 'lightningcss';

type Implementation = typeof import('lightningcss');

export type LightningCssTransformOptions = Omit<
  TransformOptions<CustomAtRules>,
  'filename' | 'code' | 'inputSourceMap'
>;

export type LightningCssLoaderOptions = LightningCssTransformOptions & {
  implementation?: Implementation;
};

export type LightningCssMinifyPluginOptions = Omit<
  LightningCssTransformOptions,
  'minify'
> & {
  implementation?: Implementation;
};

export type PluginLightningcssOptions = {
  /**
   * @see https://github.com/parcel-bundler/lightningcss/blob/master/node/index.d.ts
   * @default
   * {
   *   targets: browserslistToTargets(browserslist)
   * }
   */
  transform?: false | LightningCssLoaderOptions;
  /**
   * The 'insert-before'/'insert-after' define the insertion order of the lightningcss-loader based on the postcss-loader.
   * @default false
   */
  disableRemovePostCSS?: false | 'insert-before' | 'insert-after';
  /**
   * @see https://github.com/parcel-bundler/lightningcss/blob/master/node/index.d.ts
   * @default
   * {
   *   targets: browserslistToTargets(browserslist)
   * }
   */
  minify?: false | LightningCssMinifyPluginOptions;
};
