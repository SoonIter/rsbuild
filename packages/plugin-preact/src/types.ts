export interface PreactPluginOptions {
  jsxImportSource?: string;
}
