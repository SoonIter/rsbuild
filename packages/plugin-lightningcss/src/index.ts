export { pluginLightningcss } from './plugin';
export { default as lightningcssLoader } from './loader';
export { LightningCssMinifyPlugin } from './minimizer';
export type * from './types';

import * as lightningcss from 'lightningcss';
export { lightningcss };
