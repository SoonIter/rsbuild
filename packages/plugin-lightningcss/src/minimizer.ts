// https://github.com/fz6m/lightningcss-loader/blob/main/src/minify.ts
import { transform as _transform } from 'lightningcss';
import { Buffer } from 'node:buffer';
import type { Compilation, Compiler } from 'webpack';
import type { LightningCssMinifyPluginOptions } from './types';

const PLUGIN_NAME = 'lightningcss-minify-plugin';
const CSS_FILE_REG = /\.css(?:\?.*)?$/i;

export class LightningCssMinifyPlugin {
  private readonly options: Omit<
    LightningCssMinifyPluginOptions,
    'implementation'
  >;
  private readonly transform: typeof _transform;

  constructor(opts: LightningCssMinifyPluginOptions = {}) {
    const { implementation, ...otherOpts } = opts;
    if (implementation && typeof implementation.transform !== 'function') {
      throw new TypeError(
        `[${PLUGIN_NAME}]: implementation.transform must be an 'lightningcss' transform function. Received ${typeof implementation.transform}`,
      );
    }

    this.transform = implementation?.transform ?? _transform;
    this.options = otherOpts;
  }

  apply(compiler: Compiler) {
    compiler.hooks.compilation.tap(PLUGIN_NAME, (compilation) => {
      compilation.hooks.processAssets.tapPromise(
        {
          name: PLUGIN_NAME,
          stage: (compilation as any)?.PROCESS_ASSETS_STAGE_OPTIMIZE_SIZE,
          additionalAssets: true,
        },
        async () => await this.transformAssets(compilation),
      );

      compilation.hooks.statsPrinter.tap(PLUGIN_NAME, (statsPrinter) => {
        statsPrinter.hooks.print
          .for('asset.info.minimized')
          // @ts-ignore
          .tap(PLUGIN_NAME, (minimized, { green, formatFlag }) => {
            // @ts-ignore
            return minimized ? green(formatFlag('minimized')) : undefined;
          });
      });
    });
  }

  private async transformAssets(compilation: Compilation): Promise<void> {
    const {
      options: { devtool },
      webpack: {
        sources: { SourceMapSource, RawSource },
      },
    } = compilation.compiler;

    const sourcemap =
      this.options.sourceMap === undefined
        ? ((devtool && (devtool as string).includes('source-map')) as boolean)
        : this.options.sourceMap;

    const { targets: userTargets, ...transformOptions } = this.options;

    const assets = compilation.getAssets().filter(
      (asset) =>
        // Filter out already minimized
        !asset.info.minimized &&
        // Filter out by file type
        CSS_FILE_REG.test(asset.name),
    );

    await Promise.all(
      assets.map(async (asset) => {
        const { source, map } = asset.source.sourceAndMap();
        const sourceAsString = source.toString();
        const code = typeof source === 'string' ? Buffer.from(source) : source;

        const result = this.transform({
          filename: asset.name,
          code,
          minify: true,
          sourceMap: sourcemap,
          ...transformOptions,
        });
        const codeString = result.code.toString();

        compilation.updateAsset(
          asset.name,
          // @ts-ignore
          sourcemap
            ? new SourceMapSource(
                codeString,
                asset.name,
                JSON.parse(result.map!.toString()),
                sourceAsString,
                map as any,
                true,
              )
            : new RawSource(codeString),
          {
            ...asset.info,
            minimized: true,
          },
        );
      }),
    );
  }
}

export default LightningCssMinifyPlugin;
