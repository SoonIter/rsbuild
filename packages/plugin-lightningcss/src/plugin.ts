import {
  getBrowserslistWithDefault,
  mergeChainedOptions,
} from '@rsbuild/shared';
import type {
  ModifyBundlerChainUtils,
  RsbuildPlugin,
  BundlerChain,
} from '@rsbuild/shared';
import browserslist from '@rsbuild/shared/browserslist';
import { browserslistToTargets } from 'lightningcss';
import path from 'node:path';
import type {
  LightningCssLoaderOptions,
  PluginLightningcssOptions,
} from './types';

const PLUGIN_NAME = 'rsbuild:lightningcss';

const applyLightningcssLoader = ({
  chain,
  utils: { CHAIN_ID },
  browserslist: browserslistUserConfig,
  options,
}: {
  chain: BundlerChain;
  utils: ModifyBundlerChainUtils;
  browserslist: string[];
  options: PluginLightningcssOptions | undefined;
}) => {
  const defaultOptions = {
    targets: browserslistToTargets(browserslist(browserslistUserConfig)),
  } satisfies LightningCssLoaderOptions;

  const mergedOptions: LightningCssLoaderOptions = mergeChainedOptions({
    defaults: defaultOptions,
    options: options?.transform,
  });

  [CHAIN_ID.RULE.CSS, CHAIN_ID.RULE.SASS, CHAIN_ID.RULE.LESS].forEach(
    (ruleId) => {
      const rule = chain.module.rule(ruleId);
      const use = rule.use(CHAIN_ID.USE.LIGHTNINGCSS);

      use.loader(path.resolve(__dirname, './loader')).options(mergedOptions);

      switch (ruleId) {
        case CHAIN_ID.RULE.SASS:
          use.before(CHAIN_ID.USE.RESOLVE_URL);
          break;
        case CHAIN_ID.RULE.LESS:
          use.before(CHAIN_ID.USE.LESS);
          break;
        case CHAIN_ID.RULE.CSS:
          use.after(CHAIN_ID.USE.CSS);
          break;
      }
      rule.uses.delete(CHAIN_ID.USE.POSTCSS);
    },
  );
};

const applyLightningCssMinifyPlugin = async ({
  chain,
  utils: { CHAIN_ID },
  browserslist: browserslistUserConfig,
  options,
}: {
  chain: BundlerChain;
  browserslist: string[];
  utils: ModifyBundlerChainUtils;
  options: PluginLightningcssOptions | undefined;
}) => {
  const defaultOptions = {
    targets: browserslistToTargets(browserslist(browserslistUserConfig)),
  } satisfies LightningCssLoaderOptions;

  const mergedOptions: LightningCssLoaderOptions = mergeChainedOptions({
    defaults: defaultOptions,
    options: options?.minify,
  });

  const { LightningCssMinifyPlugin } = await import('./minimizer');
  chain.optimization
    .minimizer(CHAIN_ID.MINIMIZER.CSS)
    .use(LightningCssMinifyPlugin, [mergedOptions])
    .end();
};

export const pluginLightningcss = (
  options?: PluginLightningcssOptions,
): RsbuildPlugin => ({
  name: PLUGIN_NAME,

  setup(api) {
    api.modifyBundlerChain(async (chain, utils) => {
      const { isServer, isWebWorker, isProd, target } = utils;
      const { context } = api;
      const config = api.getNormalizedConfig();
      const browserslist = await getBrowserslistWithDefault(
        context.rootPath,
        config,
        target,
      );

      if (!isServer && !isWebWorker && options?.transform !== false) {
        applyLightningcssLoader({ chain, utils, browserslist, options });
      }

      const isMinimize = isProd && !config.output.disableMinimize;

      if (isMinimize && options?.minify !== false) {
        applyLightningCssMinifyPlugin({ chain, utils, browserslist, options });
      }
    });
  },
});
