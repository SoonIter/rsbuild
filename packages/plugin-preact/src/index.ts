import type { RsbuildPlugin } from '@rsbuild/core';
import type { PreactPluginOptions } from './types';
import { isUsingHMR } from '@rsbuild/shared';

export type PluginPreactOptions = PreactPluginOptions;

export function pluginPreact(
  _options: PluginPreactOptions = {},
): RsbuildPlugin {
  return {
    name: 'rsbuild:preact',

    pre: ['rsbuild:swc'],

    setup(api) {
      api.modifyBundlerChain(async (chain, { CHAIN_ID, isProd, target }) => {
        const rsbuildConfig = api.getNormalizedConfig();
        const usingHMR = isUsingHMR(rsbuildConfig, { isProd, target });
        const rule = chain.module.rule(CHAIN_ID.RULE.JS);

        const preactOptions = {
          development: !isProd,
          refresh: usingHMR,
          runtime: 'automatic',
          importSource: 'preact',
        };

        rule.use(CHAIN_ID.USE.SWC).tap((options) => {
          options.jsc.transform.react = {
            ...preactOptions,
          };
          return options;
        });

        if (chain.module.rules.has(CHAIN_ID.RULE.JS_DATA_URI)) {
          chain.module
            .rule(CHAIN_ID.RULE.JS_DATA_URI)
            .use(CHAIN_ID.USE.SWC)
            .tap((options) => {
              options.jsc.transform.react = {
                ...preactOptions,
              };
              return options;
            });
        }
      });
    },
  };
}
