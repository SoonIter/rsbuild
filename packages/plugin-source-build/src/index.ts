import fs from 'node:fs';
import path from 'node:path';
import { TS_CONFIG_FILE } from '@rsbuild/shared';
import type { RsbuildPlugin } from '@rsbuild/core';
import {
  filterByField,
  getDependentProjects,
  type Project,
  type ExtraMonorepoStrategies,
} from '@rsbuild/monorepo-utils';

export const PLUGIN_SOURCE_BUILD_NAME = 'rsbuild:source-build';

export const getSourceInclude = async (options: {
  projects: Project[];
  sourceField: string;
}) => {
  const { projects, sourceField } = options;

  const includes = [];
  for (const project of projects) {
    includes.push(...project.getSourceEntryPaths({ field: sourceField }));
  }

  return includes;
};

export interface PluginSourceBuildOptions {
  /**
   * Used to configure the resolve field of the source code files.
   * @default 'source''
   */
  sourceField?: string;
  /**
   * Whether to read source code or output code first.
   * @default 'source'
   */
  resolvePriority?: 'source' | 'output';
  projectName?: string;
  extraMonorepoStrategies?: ExtraMonorepoStrategies;
}

export function pluginSourceBuild(
  options?: PluginSourceBuildOptions,
): RsbuildPlugin {
  const {
    projectName,
    sourceField = 'source',
    resolvePriority = 'source',
    extraMonorepoStrategies,
  } = options ?? {};

  return {
    name: PLUGIN_SOURCE_BUILD_NAME,

    setup(api) {
      const projectRootPath = api.context.rootPath;
      const SOURCE_BUILD_RULE = 'SOURCE_BUILD_RULE';

      let projects: Project[] = [];
      let includes: string[] = [];

      api.modifyRsbuildConfig(async (config) => {
        projects = await getDependentProjects(projectName || projectRootPath, {
          cwd: projectRootPath,
          recursive: true,
          filter: filterByField(sourceField),
          extraMonorepoStrategies,
        });

        includes = await getSourceInclude({
          projects,
          sourceField,
        });

        config.source = config.source ?? {};
        config.source.include = [...(config.source.include ?? []), ...includes];
      });

      api.modifyBundlerChain((chain) => {
        const rule = chain.module.rule(SOURCE_BUILD_RULE);

        rule.include.merge([...includes, projectRootPath]);

        rule.resolve.mainFields.merge(
          resolvePriority === 'source'
            ? [sourceField, '...']
            : ['...', sourceField],
        );

        // bundler-chain do not support resolve.conditionNames yet
        rule.resolve.merge({
          // `conditionNames` is not affected by `resolvePriority`.
          // The priority is controlled by the order of fields declared in `exports`.
          conditionNames: ['...', sourceField],
        });
      });

      const getReferences = () =>
        projects
          .map((project) => path.join(project.dir, TS_CONFIG_FILE))
          .filter((filePath) => fs.existsSync(filePath));

      if (api.context.bundlerType === 'rspack') {
        api.modifyRspackConfig((config) => {
          if (!api.context.tsconfigPath) {
            return;
          }

          config.resolve ||= {};
          config.resolve.tsConfig = {
            ...config.resolve.tsConfig,
            configFile: api.context.tsconfigPath,
            references: getReferences(),
          };
        });
      } else {
        api.modifyBundlerChain((chain, { CHAIN_ID }) => {
          const { TS_CONFIG_PATHS } = CHAIN_ID.RESOLVE_PLUGIN;

          if (!chain.resolve.plugins.has(TS_CONFIG_PATHS)) {
            return;
          }

          // set references config
          // https://github.com/dividab/tsconfig-paths-webpack-plugin#options
          chain.resolve.plugin(TS_CONFIG_PATHS).tap((options) =>
            options.map((option) => ({
              ...option,
              references: getReferences(),
            })),
          );
        });
      }
    },
  };
}
