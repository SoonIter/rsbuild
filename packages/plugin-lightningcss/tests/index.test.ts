import { expect, describe, it } from 'vitest';
import { pluginLightningcss, lightningcss } from '../src';
import { createStubRsbuild } from '@scripts/test-helper';

describe('plugins/lightningcss', () => {
  it('plugin-lightningcss should replace postcss-loader with lightningcss-loader with default options', async () => {
    const rsbuild = await createStubRsbuild({
      plugins: [pluginLightningcss()],
    });

    const bundlerConfigs = await rsbuild.initConfigs();

    expect(bundlerConfigs).toMatchSnapshot();
  });

  it('plugin-lightningcss should set lightningCssMinifyPlugin with default options', async () => {
    process.env.NODE_ENV = 'production';
    const rsbuild = await createStubRsbuild({
      plugins: [pluginLightningcss()],
    });
    const config = await rsbuild.unwrapConfig();
    expect(config.optimization).toMatchSnapshot();

    process.env.NODE_ENV = 'test';
  });

  it('plugin-lightningcss should be configurable by users', async () => {
    process.env.NODE_ENV = 'production';
    const rsbuild = await createStubRsbuild({
      plugins: [
        pluginLightningcss({
          transform: {
            cssModules: {
              dashedIdents: true,
              pattern: '[hash]-[local]',
            },
            visitor: {
              Length(len) {
                return {
                  unit: 'rem',
                  value: len.value,
                };
              },
            },
          },
          minify: {
            errorRecovery: true,
            exclude: lightningcss.Features.Colors,
          },
        }),
      ],
    });

    const bundlerConfigs = await rsbuild.initConfigs();
    expect(bundlerConfigs[0]).toMatchSnapshot();

    const config = await rsbuild.unwrapConfig();
    expect(config).toMatchSnapshot();

    process.env.NODE_ENV = 'test';
  });
});
