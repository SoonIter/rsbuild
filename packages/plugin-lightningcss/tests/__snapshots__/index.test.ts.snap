// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`plugins/lightningcss > plugin-lightningcss should be configurable by users 1`] = `
{
  "module": {
    "rules": [
      {
        "use": [
          {
            "loader": "<ROOT>/packages/plugin-lightningcss/src/loader",
            "options": {
              "cssModules": {
                "dashedIdents": true,
                "pattern": "[hash]-[local]",
              },
              "targets": {
                "chrome": 5701632,
                "edge": 5767168,
                "firefox": 5111808,
                "safari": 917504,
              },
              "visitor": {
                "Length": [Function],
              },
            },
          },
        ],
      },
      {
        "use": [
          {
            "loader": "<ROOT>/packages/plugin-lightningcss/src/loader",
            "options": {
              "cssModules": {
                "dashedIdents": true,
                "pattern": "[hash]-[local]",
              },
              "targets": {
                "chrome": 5701632,
                "edge": 5767168,
                "firefox": 5111808,
                "safari": 917504,
              },
              "visitor": {
                "Length": [Function],
              },
            },
          },
        ],
      },
      {
        "use": [
          {
            "loader": "<ROOT>/packages/plugin-lightningcss/src/loader",
            "options": {
              "cssModules": {
                "dashedIdents": true,
                "pattern": "[hash]-[local]",
              },
              "targets": {
                "chrome": 5701632,
                "edge": 5767168,
                "firefox": 5111808,
                "safari": 917504,
              },
              "visitor": {
                "Length": [Function],
              },
            },
          },
        ],
      },
    ],
  },
  "optimization": {
    "minimizer": [
      LightningCssMinifyPlugin {
        "options": {
          "errorRecovery": true,
          "exclude": 64512,
          "targets": {
            "chrome": 5701632,
            "edge": 5767168,
            "firefox": 5111808,
            "safari": 917504,
          },
        },
        "transform": [Function],
      },
    ],
  },
}
`;

exports[`plugins/lightningcss > plugin-lightningcss should be configurable by users 2`] = `
{
  "module": {
    "rules": [
      {
        "use": [
          {
            "loader": "<ROOT>/packages/plugin-lightningcss/src/loader",
            "options": {
              "cssModules": {
                "dashedIdents": true,
                "pattern": "[hash]-[local]",
              },
              "targets": {
                "chrome": 5701632,
                "edge": 5767168,
                "firefox": 5111808,
                "safari": 917504,
              },
              "visitor": {
                "Length": [Function],
              },
            },
          },
        ],
      },
      {
        "use": [
          {
            "loader": "<ROOT>/packages/plugin-lightningcss/src/loader",
            "options": {
              "cssModules": {
                "dashedIdents": true,
                "pattern": "[hash]-[local]",
              },
              "targets": {
                "chrome": 5701632,
                "edge": 5767168,
                "firefox": 5111808,
                "safari": 917504,
              },
              "visitor": {
                "Length": [Function],
              },
            },
          },
        ],
      },
      {
        "use": [
          {
            "loader": "<ROOT>/packages/plugin-lightningcss/src/loader",
            "options": {
              "cssModules": {
                "dashedIdents": true,
                "pattern": "[hash]-[local]",
              },
              "targets": {
                "chrome": 5701632,
                "edge": 5767168,
                "firefox": 5111808,
                "safari": 917504,
              },
              "visitor": {
                "Length": [Function],
              },
            },
          },
        ],
      },
    ],
  },
  "optimization": {
    "minimizer": [
      LightningCssMinifyPlugin {
        "options": {
          "errorRecovery": true,
          "exclude": 64512,
          "targets": {
            "chrome": 5701632,
            "edge": 5767168,
            "firefox": 5111808,
            "safari": 917504,
          },
        },
        "transform": [Function],
      },
    ],
  },
}
`;

exports[`plugins/lightningcss > plugin-lightningcss should replace postcss-loader with lightningcss-loader with default options 1`] = `
[
  {
    "module": {
      "rules": [
        {
          "use": [
            {
              "loader": "<ROOT>/packages/plugin-lightningcss/src/loader",
              "options": {
                "targets": {
                  "chrome": 5701632,
                  "edge": 5767168,
                  "firefox": 5111808,
                  "safari": 917504,
                },
              },
            },
          ],
        },
        {
          "use": [
            {
              "loader": "<ROOT>/packages/plugin-lightningcss/src/loader",
              "options": {
                "targets": {
                  "chrome": 5701632,
                  "edge": 5767168,
                  "firefox": 5111808,
                  "safari": 917504,
                },
              },
            },
          ],
        },
        {
          "use": [
            {
              "loader": "<ROOT>/packages/plugin-lightningcss/src/loader",
              "options": {
                "targets": {
                  "chrome": 5701632,
                  "edge": 5767168,
                  "firefox": 5111808,
                  "safari": 917504,
                },
              },
            },
          ],
        },
      ],
    },
  },
]
`;

exports[`plugins/lightningcss > plugin-lightningcss should set lightningCssMinifyPlugin with default options 1`] = `
{
  "minimizer": [
    LightningCssMinifyPlugin {
      "options": {
        "targets": {
          "chrome": 5701632,
          "edge": 5767168,
          "firefox": 5111808,
          "safari": 917504,
        },
      },
      "transform": [Function],
    },
  ],
}
`;
